File: modules/kscan-pii-native/android/src/main/java/expo/modules/kscanpiinative/NativePrivacyModels.kt
package expo.modules.kscanpiinative

import android.os.Bundle
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

enum class NativePrivacyStatus(val value: String) {
    SUCCESS("success"),
    NO_FACES("no_faces"),
    UNSUPPORTED("unsupported"),
    FAILED("failed");

    companion object {
        fun from(value: String): NativePrivacyStatus? = entries.find { it.value == value }
    }
}

enum class NativePrivacyErrorCode(val value: String) {
    INVALID_INPUT("INVALID_INPUT"),
    INVALID_URI("INVALID_URI"),
    UNSUPPORTED_SCHEME("UNSUPPORTED_SCHEME"),
    UNSUPPORTED_FORMAT("UNSUPPORTED_FORMAT"),
    IMAGE_TOO_LARGE("IMAGE_TOO_LARGE"),
    DECODE_FAILED("DECODE_FAILED"),
    ORIENTATION_FAILED("ORIENTATION_FAILED"),
    DETECTOR_UNAVAILABLE("DETECTOR_UNAVAILABLE"),
    DETECTION_FAILED("DETECTION_FAILED"),
    INVALID_REGION("INVALID_REGION"),
    MASKING_FAILED("MASKING_FAILED"),
    ENCODING_FAILED("ENCODING_FAILED"),
    VERIFICATION_FAILED("VERIFICATION_FAILED"),
    CLEANUP_REJECTED("CLEANUP_REJECTED"),
    CLEANUP_FAILED("CLEANUP_FAILED"),
    INTERNAL_ERROR("INTERNAL_ERROR");

    companion object {
        fun from(value: String): NativePrivacyErrorCode? = entries.find { it.value == value }
    }
}

data class NativeFaceMaskInputRecord(
    @Field val imageUri: String = "",
    @Field val paddingRatio: Double? = null,
) : Record

data class NativeFaceMaskInput(
    val imageUri: String,
    val paddingRatio: Double = NativePrivacyConstants.DEFAULT_PADDING_RATIO,
)

data class NativeFaceRegion(
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
)

data class NativeFaceMaskResult(
    val status: NativePrivacyStatus,
    val platform: String,
    val detectorImplementation: String,
    val detectorVersion: String,
    val sanitizerVersion: String,
    val inputWidth: Int? = null,
    val inputHeight: Int? = null,
    val outputWidth: Int? = null,
    val outputHeight: Int? = null,
    val facesDetected: Int,
    val facesAccepted: Int,
    val facesMasked: Int,
    val regionsChanged: Int,
    val regionsAlreadyRedacted: Int,
    val pixelsChanged: Boolean,
    val sanitizedUri: String? = null,
    val inputChecksum: String? = null,
    val outputChecksum: String? = null,
    val checksumAlgorithm: String? = null,
    val detectionDurationMs: Long? = null,
    val maskingDurationMs: Long? = null,
    val encodingDurationMs: Long? = null,
    val verificationDurationMs: Long? = null,
    val totalDurationMs: Long? = null,
    val warnings: List<String> = emptyList(),
    val errorCode: NativePrivacyErrorCode? = null,
    val failureReason: String? = null,
) {
    fun toBundle(): Bundle = Bundle().apply {
        putString("status", status.value)
        putString("platform", platform)
        putString("detectorImplementation", detectorImplementation)
        putString("detectorVersion", detectorVersion)
        putString("sanitizerVersion", sanitizerVersion)
        inputWidth?.let { putInt("inputWidth", it) }
        inputHeight?.let { putInt("inputHeight", it) }
        outputWidth?.let { putInt("outputWidth", it) }
        outputHeight?.let { putInt("outputHeight", it) }
        putInt("facesDetected", facesDetected)
        putInt("facesAccepted", facesAccepted)
        putInt("facesMasked", facesMasked)
        putInt("regionsChanged", regionsChanged)
        putInt("regionsAlreadyRedacted", regionsAlreadyRedacted)
        putBoolean("pixelsChanged", pixelsChanged)
        sanitizedUri?.let { putString("sanitizedUri", it) }
        inputChecksum?.let { putString("inputChecksum", it) }
        outputChecksum?.let { putString("outputChecksum", it) }
        checksumAlgorithm?.let { putString("checksumAlgorithm", it) }
        detectionDurationMs?.let { putLong("detectionDurationMs", it) }
        maskingDurationMs?.let { putLong("maskingDurationMs", it) }
        encodingDurationMs?.let { putLong("encodingDurationMs", it) }
        verificationDurationMs?.let { putLong("verificationDurationMs", it) }
        totalDurationMs?.let { putLong("totalDurationMs", it) }
        putStringArray("warnings", warnings.toTypedArray())
        errorCode?.let { putString("errorCode", it.value) }
        failureReason?.let { putString("failureReason", it) }
    }
}

data class NativePrivacyCapabilities(
    val supported: Boolean,
    val platform: String,
    val detectorImplementation: String,
    val acceptedUriSchemes: List<String>,
    val acceptedMimeTypes: List<String>,
    val outputMimeType: String,
    val maxWidth: Int,
    val maxHeight: Int,
    val maxPixels: Long,
    val sanitizerVersion: String,
) {
    fun toBundle(): Bundle = Bundle().apply {
        putBoolean("supported", supported)
        putString("platform", platform)
        putString("detectorImplementation", detectorImplementation)
        putStringArray("acceptedUriSchemes", acceptedUriSchemes.toTypedArray())
        putStringArray("acceptedMimeTypes", acceptedMimeTypes.toTypedArray())
        putString("outputMimeType", outputMimeType)
        putInt("maxWidth", maxWidth)
        putInt("maxHeight", maxHeight)
        putLong("maxPixels", maxPixels)
        putString("sanitizerVersion", sanitizerVersion)
    }
}

// ── License-plate masking (Build 34, closet media privacy) ───────────────────
//
// The SECOND question the same pipeline can be asked. It shares the decoder,
// the box normalizer, the redactor, the cache manager and the output verifier
// with face masking; it differs only in which regions it selects.
//
// WHAT IS DELIBERATELY ABSENT FROM THESE MODELS: any field that could carry a
// recognized character across the bridge. There is no `text`, no `plateNumber`,
// no `sample`, no `preview`. That absence is the privacy contract, not an
// oversight — a field would eventually be filled. Counts and geometry only.

/** Mirrors NativeFaceMaskInputRecord exactly; the padding contract is shared. */
data class NativePlateMaskInputRecord(
    @Field val imageUri: String = "",
    @Field val paddingRatio: Double? = null,
) : Record

/**
 * A SEPARATE status vocabulary from NativePrivacyStatus, matching the split the
 * TypeScript contract makes (NativePlateStatus vs NativePrivacyStatus).
 *
 * The alternative — adding NO_PLATES to the shared face enum — would put a
 * value in the face path's type that the face path can never emit, and would
 * edit a vocabulary the face path already ships with. Two small enums are
 * cheaper than one enum that is a lie about one of its users.
 *
 * NO_PLATES means "the image was screened and nothing plate-shaped matched",
 * which is NOT a masked-success value: no sanitized output is produced with it.
 */
enum class NativePlateStatus(val value: String) {
    SUCCESS("success"),
    NO_PLATES("no_plates"),
    UNSUPPORTED("unsupported"),
    FAILED("failed");

    companion object {
        fun from(value: String): NativePlateStatus? = entries.find { it.value == value }
    }
}

data class NativePlateMaskResult(
    val status: NativePlateStatus,
    val platform: String,
    val detectorImplementation: String,
    val detectorVersion: String,
    val sanitizerVersion: String,
    val inputWidth: Int? = null,
    val inputHeight: Int? = null,
    val outputWidth: Int? = null,
    val outputHeight: Int? = null,
    val platesDetected: Int,
    val platesAccepted: Int,
    val platesMasked: Int,
    val regionsChanged: Int,
    val regionsAlreadyRedacted: Int,
    val pixelsChanged: Boolean,
    val sanitizedUri: String? = null,
    val inputChecksum: String? = null,
    val outputChecksum: String? = null,
    val checksumAlgorithm: String? = null,
    val detectionDurationMs: Long? = null,
    val maskingDurationMs: Long? = null,
    val encodingDurationMs: Long? = null,
    val verificationDurationMs: Long? = null,
    val totalDurationMs: Long? = null,
    val warnings: List<String> = emptyList(),
    val errorCode: NativePrivacyErrorCode? = null,
    val failureReason: String? = null,
) {
    fun toBundle(): Bundle = Bundle().apply {
        putString("status", status.value)
        putString("platform", platform)
        putString("detectorImplementation", detectorImplementation)
        putString("detectorVersion", detectorVersion)
        putString("sanitizerVersion", sanitizerVersion)
        inputWidth?.let { putInt("inputWidth", it) }
        inputHeight?.let { putInt("inputHeight", it) }
        outputWidth?.let { putInt("outputWidth", it) }
        outputHeight?.let { putInt("outputHeight", it) }
        putInt("platesDetected", platesDetected)
        putInt("platesAccepted", platesAccepted)
        putInt("platesMasked", platesMasked)
        putInt("regionsChanged", regionsChanged)
        putInt("regionsAlreadyRedacted", regionsAlreadyRedacted)
        putBoolean("pixelsChanged", pixelsChanged)
        // Present ONLY on a verified masked output. Every failure path leaves
        // it absent, so a consumer that reads this field can never be handed
        // the original image believing it was sanitized.
        sanitizedUri?.let { putString("sanitizedUri", it) }
        // Written as a LITERAL, not a constructor field, so no present or
        // future code path in this module can set it true without also
        // changing this line and the audit that reads it.
        putBoolean("ocrPerformed", false)
        inputChecksum?.let { putString("inputChecksum", it) }
        outputChecksum?.let { putString("outputChecksum", it) }
        checksumAlgorithm?.let { putString("checksumAlgorithm", it) }
        detectionDurationMs?.let { putLong("detectionDurationMs", it) }
        maskingDurationMs?.let { putLong("maskingDurationMs", it) }
        encodingDurationMs?.let { putLong("encodingDurationMs", it) }
        verificationDurationMs?.let { putLong("verificationDurationMs", it) }
        totalDurationMs?.let { putLong("totalDurationMs", it) }
        putStringArray("warnings", warnings.toTypedArray())
        errorCode?.let { putString("errorCode", it.value) }
        failureReason?.let { putString("failureReason", it) }
    }
}

/**
 * The first five fields are the shared cross-platform contract. The geometry
 * block below them is Android-specific reporting, and is here so a caller can
 * record WHICH heuristic screened a given image: a threshold change then shows
 * up in the telemetry of the build that shipped it instead of being an
 * invisible behaviour change.
 *
 * The decoder limits (accepted schemes, MIME types, max dimensions) are NOT
 * repeated here. Plate masking uses the same decoder as face masking, so
 * getPrivacyCapabilities already answers for both, and a second copy is a
 * second thing to drift.
 */
data class NativePlateCapabilities(
    val supported: Boolean,
    val platform: String,
    val detectorImplementation: String,
    val detectorVersion: String,
    val sanitizerVersion: String,
    val defaultPaddingRatio: Double,
    val minAspectRatio: Double,
    val maxAspectRatio: Double,
    val minWidthRatio: Double,
    val minHeightPx: Int,
    val minAreaRatio: Double,
) {
    fun toBundle(): Bundle = Bundle().apply {
        putBoolean("supported", supported)
        putString("platform", platform)
        putString("detectorImplementation", detectorImplementation)
        putString("detectorVersion", detectorVersion)
        putString("sanitizerVersion", sanitizerVersion)
        putDouble("defaultPaddingRatio", defaultPaddingRatio)
        putDouble("minAspectRatio", minAspectRatio)
        putDouble("maxAspectRatio", maxAspectRatio)
        putDouble("minWidthRatio", minWidthRatio)
        putInt("minHeightPx", minHeightPx)
        putDouble("minAreaRatio", minAreaRatio)
        // Literal, not a field: the contract's auditable claim that no
        // character recognition is performed on any path. Nothing in this
        // module can set it true without editing this line.
        putBoolean("ocrPerformed", false)
    }
}

data class NativeCleanupResult(
    val deleted: Boolean,
    val rejected: Boolean,
    val warnings: List<String> = emptyList(),
) {
    fun toBundle(): Bundle = Bundle().apply {
        putBoolean("deleted", deleted)
        putBoolean("rejected", rejected)
        putStringArray("warnings", warnings.toTypedArray())
    }
}

File: modules/kscan-pii-native/ios/NativePrivacyModels.swift
import Foundation

enum NativePrivacyStatus: String {
    case success = "success"
    case noFaces = "no_faces"
    case unsupported = "unsupported"
    case failed = "failed"
}

enum NativePrivacyErrorCode: String {
    case invalidInput = "INVALID_INPUT"
    case invalidUri = "INVALID_URI"
    case unsupportedScheme = "UNSUPPORTED_SCHEME"
    case unsupportedFormat = "UNSUPPORTED_FORMAT"
    case imageTooLarge = "IMAGE_TOO_LARGE"
    case decodeFailed = "DECODE_FAILED"
    case orientationFailed = "ORIENTATION_FAILED"
    case detectorUnavailable = "DETECTOR_UNAVAILABLE"
    case detectionFailed = "DETECTION_FAILED"
    case invalidRegion = "INVALID_REGION"
    case maskingFailed = "MASKING_FAILED"
    case encodingFailed = "ENCODING_FAILED"
    case verificationFailed = "VERIFICATION_FAILED"
    case cleanupRejected = "CLEANUP_REJECTED"
    case cleanupFailed = "CLEANUP_FAILED"
    case internalError = "INTERNAL_ERROR"
}

struct NativeFaceMaskInput {
    let imageUri: String
    let paddingRatio: Double
}

struct NativeFaceRegion {
    let x: Int
    let y: Int
    let width: Int
    let height: Int
}

struct NativeFaceMaskResult {
    let status: NativePrivacyStatus
    let platform: String
    let detectorImplementation: String
    let detectorVersion: String
    let sanitizerVersion: String
    let inputWidth: Int?
    let inputHeight: Int?
    let outputWidth: Int?
    let outputHeight: Int?
    let facesDetected: Int
    let facesAccepted: Int
    let facesMasked: Int
    let regionsChanged: Int
    let regionsAlreadyRedacted: Int
    let pixelsChanged: Bool
    let sanitizedUri: String?
    let inputChecksum: String?
    let outputChecksum: String?
    let checksumAlgorithm: String?
    let detectionDurationMs: Int?
    let maskingDurationMs: Int?
    let encodingDurationMs: Int?
    let verificationDurationMs: Int?
    let totalDurationMs: Int?
    let warnings: [String]
    let errorCode: NativePrivacyErrorCode?
    let failureReason: String?

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "status": status.rawValue,
            "platform": platform,
            "detectorImplementation": detectorImplementation,
            "detectorVersion": detectorVersion,
            "sanitizerVersion": sanitizerVersion,
            "facesDetected": facesDetected,
            "facesAccepted": facesAccepted,
            "facesMasked": facesMasked,
            "regionsChanged": regionsChanged,
            "regionsAlreadyRedacted": regionsAlreadyRedacted,
            "pixelsChanged": pixelsChanged,
            "warnings": warnings,
        ]
        if let inputWidth = inputWidth { dict["inputWidth"] = inputWidth }
        if let inputHeight = inputHeight { dict["inputHeight"] = inputHeight }
        if let outputWidth = outputWidth { dict["outputWidth"] = outputWidth }
        if let outputHeight = outputHeight { dict["outputHeight"] = outputHeight }
        if let sanitizedUri = sanitizedUri { dict["sanitizedUri"] = sanitizedUri }
        if let inputChecksum = inputChecksum { dict["inputChecksum"] = inputChecksum }
        if let outputChecksum = outputChecksum { dict["outputChecksum"] = outputChecksum }
        if let checksumAlgorithm = checksumAlgorithm { dict["checksumAlgorithm"] = checksumAlgorithm }
        if let detectionDurationMs = detectionDurationMs { dict["detectionDurationMs"] = detectionDurationMs }
        if let maskingDurationMs = maskingDurationMs { dict["maskingDurationMs"] = maskingDurationMs }
        if let encodingDurationMs = encodingDurationMs { dict["encodingDurationMs"] = encodingDurationMs }
        if let verificationDurationMs = verificationDurationMs { dict["verificationDurationMs"] = verificationDurationMs }
        if let totalDurationMs = totalDurationMs { dict["totalDurationMs"] = totalDurationMs }
        if let errorCode = errorCode { dict["errorCode"] = errorCode.rawValue }
        if let failureReason = failureReason { dict["failureReason"] = failureReason }
        return dict
    }
}

struct NativePrivacyCapabilities {
    let supported: Bool
    let platform: String
    let detectorImplementation: String
    let acceptedUriSchemes: [String]
    let acceptedMimeTypes: [String]
    let outputMimeType: String
    let maxWidth: Int
    let maxHeight: Int
    let maxPixels: Int
    let sanitizerVersion: String

    func toDictionary() -> [String: Any] {
        return [
            "supported": supported,
            "platform": platform,
            "detectorImplementation": detectorImplementation,
            "acceptedUriSchemes": acceptedUriSchemes,
            "acceptedMimeTypes": acceptedMimeTypes,
            "outputMimeType": outputMimeType,
            "maxWidth": maxWidth,
            "maxHeight": maxHeight,
            "maxPixels": maxPixels,
            "sanitizerVersion": sanitizerVersion,
        ]
    }
}

// ── License-plate REGION screening ───────────────────────────────────────────
//
// A separate status enum rather than a new case on NativePrivacyStatus, and a
// separate result struct rather than a reuse of NativeFaceMaskResult. Both
// follow the precedent the person-detection capability already set with
// NativeExtractionStatus.
//
// The result struct is the load-bearing one. NativeFaceMaskResult's counters are
// literally named facesDetected / facesAccepted / facesMasked, and they cross the
// bridge under those keys. Reporting three masked plates as `facesDetected: 3`
// would be a lie told by a privacy engine about what it did — and the JS layer
// this feeds (services/privacyImageUpload.ts) reports faceMaskApplied and
// plateMaskApplied as two independent claims, so it needs two independent
// answers. The SHAPE is mirrored field for field on purpose; only the three
// counters, the status type and the version strings differ.

enum NativePlateStatus: String {
    case success = "success"
    /// "No region matched the plate GEOMETRY" — never "this image contains no
    /// plate". See the limitations in IOSPlateDetector's header.
    case noPlates = "no_plates"
    case unsupported = "unsupported"
    case failed = "failed"
}

struct NativePlateMaskInput {
    let imageUri: String
    let paddingRatio: Double
}

struct NativePlateMaskResult {
    let status: NativePlateStatus
    let platform: String
    let detectorImplementation: String
    let detectorVersion: String
    let sanitizerVersion: String
    /// Always false on this path. Named to match the shared cross-platform
    /// contract (recognizedTextConsumed, not ocrPerformed): on iOS
    /// VNDetectTextRectanglesRequest never performs character recognition at
    /// all, so both readings of the claim are true here, but the field name
    /// must agree with Android, where recognition DOES run internally and
    /// only "the text was never consumed" is the honest claim. Emitted rather
    /// than assumed so that a caller making a privacy claim reads it from the
    /// result instead of from a comment. See IOSPlateDetector's header.
    let recognizedTextConsumed: Bool
    let inputWidth: Int?
    let inputHeight: Int?
    let outputWidth: Int?
    let outputHeight: Int?
    let platesDetected: Int
    let platesAccepted: Int
    let platesMasked: Int
    let regionsChanged: Int
    let regionsAlreadyRedacted: Int
    let pixelsChanged: Bool
    let sanitizedUri: String?
    let inputChecksum: String?
    let outputChecksum: String?
    let checksumAlgorithm: String?
    let detectionDurationMs: Int?
    let maskingDurationMs: Int?
    let encodingDurationMs: Int?
    let verificationDurationMs: Int?
    let totalDurationMs: Int?
    let warnings: [String]
    let errorCode: NativePrivacyErrorCode?
    let failureReason: String?

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "status": status.rawValue,
            "platform": platform,
            "detectorImplementation": detectorImplementation,
            "detectorVersion": detectorVersion,
            "sanitizerVersion": sanitizerVersion,
            "recognizedTextConsumed": recognizedTextConsumed,
            "platesDetected": platesDetected,
            "platesAccepted": platesAccepted,
            "platesMasked": platesMasked,
            "regionsChanged": regionsChanged,
            "regionsAlreadyRedacted": regionsAlreadyRedacted,
            "pixelsChanged": pixelsChanged,
            "warnings": warnings,
        ]
        if let inputWidth = inputWidth { dict["inputWidth"] = inputWidth }
        if let inputHeight = inputHeight { dict["inputHeight"] = inputHeight }
        if let outputWidth = outputWidth { dict["outputWidth"] = outputWidth }
        if let outputHeight = outputHeight { dict["outputHeight"] = outputHeight }
        if let sanitizedUri = sanitizedUri { dict["sanitizedUri"] = sanitizedUri }
        if let inputChecksum = inputChecksum { dict["inputChecksum"] = inputChecksum }
        if let outputChecksum = outputChecksum { dict["outputChecksum"] = outputChecksum }
        if let checksumAlgorithm = checksumAlgorithm { dict["checksumAlgorithm"] = checksumAlgorithm }
        if let detectionDurationMs = detectionDurationMs { dict["detectionDurationMs"] = detectionDurationMs }
        if let maskingDurationMs = maskingDurationMs { dict["maskingDurationMs"] = maskingDurationMs }
        if let encodingDurationMs = encodingDurationMs { dict["encodingDurationMs"] = encodingDurationMs }
        if let verificationDurationMs = verificationDurationMs { dict["verificationDurationMs"] = verificationDurationMs }
        if let totalDurationMs = totalDurationMs { dict["totalDurationMs"] = totalDurationMs }
        if let errorCode = errorCode { dict["errorCode"] = errorCode.rawValue }
        if let failureReason = failureReason { dict["failureReason"] = failureReason }
        return dict
    }
}

struct NativePlateCapabilities {
    let supported: Bool
    let platform: String
    let detectorImplementation: String
    let detectorVersion: String
    let sanitizerVersion: String
    /// Same cross-platform field as NativePlateMaskResult; see its comment.
    let recognizedTextConsumed: Bool
    let acceptedUriSchemes: [String]
    let acceptedMimeTypes: [String]
    let outputMimeType: String
    let maxWidth: Int
    let maxHeight: Int
    let maxPixels: Int
    let defaultPaddingRatio: Double
    let minPaddingRatio: Double
    let maxPaddingRatio: Double

    /**
     The geometry thresholds are reported, not just applied.

     A caller deciding how much weight to put on a `no_plates` result needs to
     know what "plate-like" meant to the build it is talking to. Without this the
     only way to find out is to read the Swift, which a QA pass on a shipped
     binary cannot do. It also gives the shared test suite something to pin, so a
     threshold cannot be retuned without the change being visible.
     */
    func toDictionary() -> [String: Any] {
        // Annotated, and built as its own binding rather than inlined below. The
        // ratios are Double and the pixel floors are Int, so as an inline nested
        // literal this is a heterogeneous collection Swift refuses to infer.
        let geometryThresholds: [String: Any] = [
            "minAspectRatio": NativePrivacyConstants.plateMinAspectRatio,
            "maxAspectRatio": NativePrivacyConstants.plateMaxAspectRatio,
            "minRelativeWidth": NativePrivacyConstants.plateMinRelativeWidth,
            "minRelativeArea": NativePrivacyConstants.plateMinRelativeArea,
            "minPixelWidth": NativePrivacyConstants.plateMinPixelWidth,
            "minPixelHeight": NativePrivacyConstants.plateMinPixelHeight,
        ]

        return [
            "supported": supported,
            "platform": platform,
            "detectorImplementation": detectorImplementation,
            "detectorVersion": detectorVersion,
            "sanitizerVersion": sanitizerVersion,
            "recognizedTextConsumed": recognizedTextConsumed,
            "acceptedUriSchemes": acceptedUriSchemes,
            "acceptedMimeTypes": acceptedMimeTypes,
            "outputMimeType": outputMimeType,
            "maxWidth": maxWidth,
            "maxHeight": maxHeight,
            "maxPixels": maxPixels,
            "defaultPaddingRatio": defaultPaddingRatio,
            "minPaddingRatio": minPaddingRatio,
            "maxPaddingRatio": maxPaddingRatio,
            "geometryThresholds": geometryThresholds,
        ]
    }
}

struct NativeCleanupResult {
    let deleted: Bool
    let rejected: Bool
    let warnings: [String]

    func toDictionary() -> [String: Any] {
        return [
            "deleted": deleted,
            "rejected": rejected,
            "warnings": warnings,
        ]
    }
}
